File: src/Core.CSharp/ZSetMerkle.cs
using System;
using System.Buffers.Binary;
using System.Collections.Immutable;
using System.IO.Hashing;
using System.Runtime.InteropServices;

namespace Zeta.Core.CSharp;

/// <summary>
/// Canonical Merkle root over a Z-set — the C# parity oracle for <c>src/Core/ZSetMerkle.fs</c>.
/// Leaves are the DBSP Z-set entries, so the content-addressed root is a pure function of the net Z-set state.
/// </summary>
public static class ZSetMerkle
{
    /// <summary>
    /// Canonical leaf encoding for one <c>(key, weight)</c> entry:
    /// <c>[4-byte LE keyLen][keyBytes][8-byte LE weight]</c>.
    /// </summary>
    private static byte[] LeafBytes(byte[] keyBytes, long weight)
    {
        var buf = new byte[4 + keyBytes.Length + 8];
        BinaryPrimitives.WriteInt32LittleEndian(buf.AsSpan(0, 4), keyBytes.Length);
        keyBytes.CopyTo(buf, 4);
        BinaryPrimitives.WriteInt64LittleEndian(buf.AsSpan(4 + keyBytes.Length, 8), weight);
        return buf;
    }

    /// <summary>
    /// Combine two child digests into a parent: 32 LE bytes <c>a.Hi a.Lo b.Hi b.Lo</c>, re-hashed.
    /// </summary>
    private static MerkleHash Combine(Func<byte[], MerkleHash> hash, MerkleHash a, MerkleHash b)
    {
        var buf = new byte[32];
        BinaryPrimitives.WriteUInt64LittleEndian(buf.AsSpan(0, 8), a.Hi);
        BinaryPrimitives.WriteUInt64LittleEndian(buf.AsSpan(8, 8), a.Lo);
        BinaryPrimitives.WriteUInt64LittleEndian(buf.AsSpan(16, 8), b.Hi);
        BinaryPrimitives.WriteUInt64LittleEndian(buf.AsSpan(24, 8), b.Lo);
        return hash(buf);
    }

    /// <summary>
    /// Fold a level of digests bottom-up; an odd trailing node is promoted (duplicated) — the standard Merkle construction.
    /// </summary>
    private static MerkleHash Fold(Func<byte[], MerkleHash> hash, MerkleHash[] level)
    {
        if (level.Length == 0)
        {
            return hash(Array.Empty<byte>());
        }

        var cur = level;
        while (cur.Length > 1)
        {
            var parent = new MerkleHash[(cur.Length + 1) / 2];
            for (var i = 0; i < parent.Length; i++)
            {
                var left = cur[2 * i];
                var right = 2 * i + 1 < cur.Length ? cur[2 * i + 1] : left;
                parent[i] = Combine(hash, left, right);
            }
            cur = parent;
        }

        return cur[0];
    }

    /// <summary>
    /// Canonical Merkle root over <c>z</c> with an explicit hash function. Leaves = <c>(key, weight)</c> entries
    /// encoded + sorted by key bytes (ordinal); folded bottom-up. Deterministic + retraction-native.
    /// </summary>
    public static MerkleHash RootWith<T>(Func<byte[], MerkleHash> hash, Func<T, byte[]> encodeKey, ZSet<T> z)
    {
        ArgumentNullException.ThrowIfNull(hash);
        ArgumentNullException.ThrowIfNull(encodeKey);
        ArgumentNullException.ThrowIfNull(z);

        var items = z.ToImmutableArray();
        var leavesTemp = new (byte[] KeyBytes, long Weight)[items.Length];
        for (var i = 0; i < items.Length; i++)
        {
            leavesTemp[i] = (encodeKey(items[i].Key), items[i].Weight);
        }

        // Lexicographic ordinal comparison of key byte arrays (the cross-language canonical order)
        Array.Sort(leavesTemp, (a, b) => a.KeyBytes.AsSpan().SequenceCompareTo(b.KeyBytes.AsSpan()));

        var leaves = new MerkleHash[leavesTemp.Length];
        for (var i = 0; i < leavesTemp.Length; i++)
        {
            leaves[i] = hash(LeafBytes(leavesTemp[i].KeyBytes, leavesTemp[i].Weight));
        }

        return Fold(hash, leaves);
    }

    /// <summary>
    /// Canonical Merkle root using the default digest (XxHash128 via <see cref="MerkleHash"/>).
    /// </summary>
    public static MerkleHash Root<T>(Func<T, byte[]> encodeKey, ZSet<T> z)
    {
        return RootWith(DefaultHash, encodeKey, z);
    }

    private static MerkleHash DefaultHash(byte[] bytes)
    {
        Span<byte> buf = stackalloc byte[16];
        XxHash128.Hash(bytes, buf);
        var lo = BinaryPrimitives.ReadUInt64LittleEndian(buf.Slice(0, 8));
        var hi = BinaryPrimitives.ReadUInt64LittleEndian(buf.Slice(8, 8));
        return new MerkleHash(hi, lo);
    }
}

File: src/Core.TypeScript/z-set-merkle/z-set-merkle.js
import { ofBytes } from "../merkle/merkle";
/** Lexicographic ordinal comparison of two byte arrays. */
function byteCompare(a, b) {
    const n = Math.min(a.length, b.length);
    for (let i = 0; i < n; i++) {
        if (a[i] !== b[i]) {
            return a[i] - b[i];
        }
    }
    return a.length - b.length;
}
/** Canonical leaf encoding for one entry: [4-byte LE keyLen][keyBytes][8-byte LE weight]. */
function leafBytes(keyBytes, weight) {
    const buf = new Uint8Array(4 + keyBytes.length + 8);
    const view = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);
    view.setInt32(0, keyBytes.length, true);
    buf.set(keyBytes, 4);
    view.setBigInt64(4 + keyBytes.length, weight, true);
    return buf;
}
/** Combine two child digests into a parent: 32 LE bytes a.Hi a.Lo b.Hi b.Lo, re-hashed. */
function combine(hash, a, b) {
    const buf = new Uint8Array(32);
    const view = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);
    view.setBigUint64(0, a.hi, true);
    view.setBigUint64(8, a.lo, true);
    view.setBigUint64(16, b.hi, true);
    view.setBigUint64(24, b.lo, true);
    return hash(buf);
}
/** Fold a level of digests bottom-up; an odd trailing node is promoted (duplicated). */
function fold(hash, level) {
    if (level.length === 0) {
        return hash(new Uint8Array(0));
    }
    let cur = level;
    while (cur.length > 1) {
        const parent = [];
        for (let i = 0; i < cur.length; i += 2) {
            const left = cur[i];
            const right = i + 1 < cur.length ? cur[i + 1] : left;
            parent.push(combine(hash, left, right));
        }
        cur = parent;
    }
    return cur[0];
}
/**
 * Canonical Merkle root over z with an explicit hash function. Leaves = (key, weight) entries
 * encoded + sorted by key bytes (ordinal); folded bottom-up. Deterministic + retraction-native.
 */
export function rootWith(hash, encodeKey, z) {
    const leavesTemp = z.map((e) => ({
        kb: encodeKey(e.e),
        w: e.w,
    }));
    leavesTemp.sort((a, b) => byteCompare(a.kb, b.kb));
    const leaves = leavesTemp.map((e) => hash(leafBytes(e.kb, BigInt(e.w))));
    return fold(hash, leaves);
}
/**
 * Canonical Merkle root using the default digest (XxHash128 via MerkleHash).
 */
export function root(encodeKey, z) {
    return rootWith(ofBytes, encodeKey, z);
}
